dask/dataframe/io/parquet/core.py

```diff
@@ -0,0 +1,472 @@
+        to compressors like ``{"name": "gzip", "values": "snappy"}``. The
+        default is ``"default"``, which uses the default compression for
+        whichever engine is selected.
+    write_index : boolean, optional
+        Whether or not to write the index. Defaults to True *if* divisions are
+        known.
+    append : bool, optional
+        If False (default), construct data-set from scratch. If True, add new
+        row-group(s) to an existing data-set. In the latter case, the data-set
+        must exist, and the schema must match the input data.
+    partition_on : list, optional
+        Construct directory-based partitioning by splitting on these fields'
+        values. Each dask partition will result in one or more datafiles,
+        there will be no global groupby.
+    storage_options : dict, optional
+        Key/value pairs to be passed on to the file-system backend, if any.
+    write_metadata_file : bool, optional
+        Whether to create the special "_metadata" file,
+    compute : bool, optional
+        If True (default) then the result is computed immediately. If False
+        then a ``dask.delayed`` object is returned for future computation.
+    **kwargs
+        Extra options to be passed on to the specific backend.
+
+    Examples
+    --------
+    >>> df = dd.read_csv(...)  # doctest: +SKIP
+    >>> to_parquet('/path/to/output/', df, compression='snappy')  # doctest: +SKIP
+
+    See Also
+    --------
+    read_parquet: Read parquet data to dask.dataframe
+    """
+    partition_on = partition_on or []
+    if isinstance(partition_on, string_types):
+        partition_on = [partition_on]
+
+    if set(partition_on) - set(df.columns):
+        raise ValueError(
+            "Partitioning on non-existent column. "
+            "partition_on=%s ."
+            "columns=%s" % (str(partition_on), str(list(df.columns)))
+        )
+
+    if compression != "default":
+        kwargs["compression"] = compression
+    elif "snappy" in compress:
+        kwargs["compression"] = "snappy"
+
+    if isinstance(engine, str):
+        engine = get_engine(engine)
+
+    fs, fs_token, _ = get_fs_token_paths(
+        path, mode="wb", storage_options=storage_options
+    )
+    # Trim any protocol information from the path before forwarding
+    # ideally, this should be done as a method of the file-system
+    path = infer_storage_options(path)["path"]
+
+    if write_index:
+        df = df.reset_index()
+
+    # create parquet metadata, includes loading of existing stuff is appending
+    meta, filenames, kwargs2 = engine.create_metadata(df, fs, path, append)
+    kwargs.update(kwargs2)
+
+    # write parts
+    dwrite = delayed(engine.write)
+    parts = [dwrite(
+        d, path, fs, filename, partition_on, write_metadata_file, **kwargs)
+        for d, filename in zip(df.to_delayed(), filenames)]
+
+    # single task to complete
+    out = engine.write_metadata(parts, meta, fs, path, **kwargs2)
```

Comment:
The current implementation in the FastParquet engine seems to be concrete.  Do we mean to wrap this in a delayed call?

Reply:
Yes, I think so. I'll go over this again yet and get at least the FP arm working.